stdlib/Random/src/Xoshiro.jl

```diff
@@ -113,12 +113,19 @@ struct TaskLocalRNG <: AbstractRNG end
 TaskLocalRNG(::Nothing) = TaskLocalRNG()
 rng_native_52(::TaskLocalRNG) = UInt64
 
-function setstate!(x::TaskLocalRNG, s0::UInt64, s1::UInt64, s2::UInt64, s3::UInt64)
+function setstate!(
+    x::TaskLocalRNG,
+    s0::UInt64, s1::UInt64, s2::UInt64, s3::UInt64, # xoshiro256 state
+    s4::UInt64 = hash((s0, s1)), # splitmix weight rng state
```

Comment:
Should this just multiply too, like the split, or is the slightly more complex multiply hidden in the hash function preferred?

Reply:
Keep in mind that we don't care about speed of this—we're already doing a SHA256 hash operation to generate seed values in the common path. This only happens when the user explicitly calls `Random.seed!`, which is apparently rare enough that no one has complained since we made it much slower (and safer), so I think doing a little extra work here to make these values well scrambled is ok.

Reply:
multiply is fairly close to perfect bitmixing.

Reply:
Okay, I checked the assembly, and looks like this `hash` calls does some variety of bitmixing instead of a multiply (it actually is an add between them and a constant after shifting some bits–the same in each):
```
  %5 = bitcast [2 x i64]* %0 to <2 x i64>*
  %6 = load <2 x i64>, <2 x i64>* %5, align 8
  %7 = xor <2 x i64> %6, <i64 -1, i64 -1>
  %8 = shl <2 x i64> %6, <i64 21, i64 21>
  %9 = add <2 x i64> %8, %7
  %10 = lshr <2 x i64> %9, <i64 24, i64 24>
  %11 = xor <2 x i64> %10, %9
  %12 = mul <2 x i64> %11, <i64 265, i64 265>
  %13 = lshr <2 x i64> %12, <i64 14, i64 14>
  %14 = xor <2 x i64> %13, %12
  %15 = mul <2 x i64> %14, <i64 21, i64 21>
  %16 = lshr <2 x i64> %15, <i64 28, i64 28>
  %17 = xor <2 x i64> %16, %15
  %18 = mul <2 x i64> %17, <i64 2147483649, i64 -6442450947>
  %19 = extractelement <2 x i64> %18, i64 1
  %.neg6 = add i64 %19, 3912697229818339088
  %20 = extractelement <2 x i64> %18, i64 0
  %21 = add i64 %.neg6, %20
  ret i64 %21
```